esp-hal-common/src/rtc_cntl.rs

```diff
@@ -1,46 +1,600 @@
-use crate::pac::RTC_CNTL;
+use fugit::{MegahertzU32, MicrosDurationU64};
 
-pub struct RtcCntl {
-    rtc_cntl: RTC_CNTL,
+use embedded_hal::watchdog::{Watchdog, WatchdogDisable, WatchdogEnable};
+
+use crate::{clock::Clock, clock::XtalClock, pac::RTC_CNTL, pac::TIMG0};
+
+#[cfg(not(feature = "esp32"))]
+use crate::efuse::Efuse;
+
+use crate::rom::esp_rom_delay_us;
+
+#[cfg_attr(feature = "esp32", path = "rtc/esp32.rs")]
+#[cfg_attr(feature = "esp32s2", path = "rtc/esp32s2.rs")]
+#[cfg_attr(feature = "esp32s3", path = "rtc/esp32s3.rs")]
+#[cfg_attr(feature = "esp32c3", path = "rtc/esp32c3.rs")]
+mod rtc;
+
+#[allow(unused)]
+#[derive(Debug, Clone, Copy)]
+/// RTC SLOW_CLK frequency values
+pub(crate) enum RtcFastClock {
+    /// Main XTAL, divided by 4
+    RtcFastClockXtalD4 = 0,
+    /// Internal 8 MHz RC oscillator
+    RtcFastClock8m = 1,
 }
 
-impl RtcCntl {
+impl Clock for RtcFastClock {
+    fn frequency(&self) -> MegahertzU32 {
+        match self {
+            RtcFastClock::RtcFastClockXtalD4 => MegahertzU32::Hz(40_000_000 / 4),
+            RtcFastClock::RtcFastClock8m => MegahertzU32::Hz(8_500_000),
+        }
+    }
+}
+
+#[allow(unused)]
+#[derive(Debug, Clone, Copy)]
+/// RTC SLOW_CLK frequency values
+pub(crate) enum RtcSlowClock {
+    /// Internal 150 KHz RC oscillator
+    RtcSlowClockRtc = 0,
+    /// External 32 KHz XTAL
+    RtcSlowClock32kXtal = 1,
+    /// Internal 8 MHz RC oscillator, divided by 256
+    RtcSlowClock8mD256 = 2,
+}
+
+impl Clock for RtcSlowClock {
+    fn frequency(&self) -> MegahertzU32 {
+        match self {
+            RtcSlowClock::RtcSlowClockRtc => MegahertzU32::Hz(1_500_000),
+            RtcSlowClock::RtcSlowClock32kXtal => MegahertzU32::Hz(32768),
+            RtcSlowClock::RtcSlowClock8mD256 => MegahertzU32::Hz(8_500_000 / 256),
+        }
+    }
+}
+
+#[allow(unused)]
+#[derive(Debug, Clone, Copy)]
+/// Clock source to be calibrated using rtc_clk_cal function
+pub(crate) enum RtcCalSel {
+    /// Currently selected RTC SLOW_CLK
+    RtcCalRtcMux = 0,
+    /// Internal 8 MHz RC oscillator, divided by 256
+    RtcCal8mD256 = 1,
+    /// External 32 KHz XTAL
+    RtcCal32kXtal = 2,
+    #[cfg(not(feature = "esp32"))]
+    /// Internal 150 KHz RC oscillator
+    RtcCalInternalOsc = 3,
+}
+
+pub struct Rtc {
```

Comment:
Is there a particular reason for the rename? It seems most trms refer to this block as "Low power management (RTC_CNTL)".

Reply:
Oh actually the esp32 trm states it as RTC in the glossary

![image](https://user-images.githubusercontent.com/6977289/182953094-1c52fc63-3474-4673-b8c9-c22a60f390e8.png)